File: 1021-distribute-coins-in-binary-tree/distribute-coins-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans=0;
    int distribute(TreeNode root){
        if(root==null) return 0;
        int leftCoins=distribute(root.left);
        int rightCoins=distribute(root.right);
        int total=leftCoins+rightCoins;
        root.val=root.val+total;
        ans+=Math.abs(root.val-1);
        return root.val-1;
    }
    public int distributeCoins(TreeNode root) {
        ans=0;
        distribute(root);
        return ans;
    }
}

File: 1021-distribute-coins-in-binary-tree/distribute-coins-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int distributeCoins(TreeNode* root) {
        function<int(TreeNode*,int&)> solve=[&](TreeNode* root,int& ans)->int{
            if(root==nullptr) return 0;
            int left=solve(root->left,ans);
            int right=solve(root->right,ans);
            int req_or_need=root->val+left+right-1;
            ans+=abs(req_or_need);
            return req_or_need;
        };
        int ans=0;
        solve(root,ans);
        return ans;
    }
};